thevoid/server.cpp

```diff
@@ -50,9 +50,62 @@
 namespace ioremap {
 namespace thevoid {
 
-class server_data;
+namespace signal_handler {
 
-static std::weak_ptr<signal_handler> global_signal_set;
+std::mutex lock;
+std::set<base_server*> all_servers;
+
+static
+void stop(int signal_value)
+{
+	std::lock_guard<std::mutex> locker(lock);
```

Comment:
Danil, try to implement handling of signals using signalfd, you may listen it at monitoring thread. As it's a common file descriptor you'll be allows to do whatever you want including using of mutexes and writing to logger.